File: bcs-cc/components/bcs/mesos.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
 * Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package bcs

import (
	"encoding/json"
	"fmt"
	"time"

	"bcs_cc/config"
	"bcs_cc/logging"
	"bcs_cc/utils"
)

// agent info, include ip, etc,
type agentInfo struct {
	IP       string `json:"ip"`
	Disabled bool   `json:"disabled"`
}

// reserve metric info for mesos
type bcsReponseData struct {
	TotalDisk float64     `json:"disktotal"`
	UsedDisk  float64     `json:"diskused"`
	TotalMem  float64     `json:"memtotal"`
	UsedMem   float64     `json:"memused"`
	TotalCPU  float64     `json:"cputotal"`
	UsedCPU   float64     `json:"cpuused"`
	Agents    []agentInfo `json:"agents"`
}

type bcsResponse struct {
	Code    int            `json:"code"`
	Message string         `json:"message"`
	Data    bcsReponseData `json:"data"`
}

// MesosIPResource : get mesos resource info, in order to add or update node record
func MesosIPResource(clusterID string, env string, accessToken string) ([]map[string]string, error) {
	confBase := config.GlobalConfigurations
	reqURL := fmt.Sprintf("%s/%s%s", confBase.APIGWConf.Host, env, confBase.BCSConf.MesosResourcePath)
	params := fmt.Sprintf("access_token=%s", accessToken)
	// compose the request
	req := utils.GoReq{
		Method:  "GET",
		URL:     reqURL,
		Params:  params,
		Data:    make(map[string]interface{}),
		Header:  map[string]string{"Content-Type": "application/json", "BCS-ClusterID": clusterID},
		Timeout: time.Duration(10),
	}
	body, err := utils.RequestClient(req, confBase.APIGWConf.Proxy)
	if err != nil {
		logging.Error("request bcs api error，url: %v, detail: %v", reqURL, err)
		return nil, err
	}
	var bcsResp bcsResponse
	if err := json.Unmarshal([]byte(body), &bcsResp); err != nil {
		logging.Error("parse bcs response error, cluster_id: %s, detial: %v", clusterID, err)
		return nil, err
	}
	if bcsResp.Code != 0 {
		logging.Error("bcs api response error，cluster_id: %s, url: %v, detail: %v", clusterID, reqURL, bcsResp.Message)
		return nil, fmt.Errorf("request bcs error")
	}
	var nodeList []map[string]string
	// mesos status is normal if ip in agent
	for _, agent := range bcsResp.Data.Agents {
		status := utils.Normal
		if agent.Disabled {
			status = utils.ToRemoved
		}
		nodeList = append(nodeList, map[string]string{"ip": agent.IP, "status": status})
	}

	return nodeList, nil
}

File: bcs-app/backend/container_service/clusters/flow_views/configs/mesos.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

MESOS 获取相关配置
"""
import base64
import json
import time

from django.utils.translation import ugettext_lazy as _

from backend.components import paas_cc
from backend.container_service.clusters.models import CommonStatus
from backend.utils.errcodes import ErrorCode
from backend.utils.error_codes import error_codes


class BaseConfig(object):
    def get_config(self, cluster_id, master_ips, node_ip_list=[]):
        """渲染集群或节点配置"""
        pass


class ClusterConfig(BaseConfig):
    def __init__(self, base_cluster_config, area_info, cluster_name=""):
        self.mesos_config = base_cluster_config
        self.area_config = json.loads(area_info.get('configuration', '{}'))
        self.cluster_name = cluster_name

    def get_request_config(self, cluster_id, master_ips, need_nat=True, **kwargs):
        self.mesos_config.update(
            {
                "master_iplist": ",".join(master_ips),
                "cluster_id": cluster_id,
                "check_iptables_nat": "Y" if need_nat else "N",
            }
        )
        return self.mesos_config


class NodeConfig(BaseConfig):
    def __init__(self, snapshot_config, op_type=None):
        self.mesos_config = snapshot_config

    def get_request_config(self, access_token, project_id, cluster_id, master_ip_list, ip_list):
        # TODO：现阶段slave安装和master安装配置一样，后续再增加其它
        return self.mesos_config
